piping error codes (#365)
ros-planning/navigation.ros.org: tutorials/docs/adding_a_nav2_task_server.rst

1. Add a new ``ComposableNode()`` instance in your launch file pointing to the component container of your choice.

    .. code-block:: python

        container = ComposableNodeContainer(
            name='my_container',
            namespace='',
            package='rclcpp_components',
            executable='component_container',
            composable_node_descriptions=[
                ComposableNode(
                    package='nav2_route_server',
                    plugin='nav2_route_server::RouteServer',
                    name='nav2_route_server'),
            ],
            output='screen',
        )

    .. seealso::
        See example in composition demo's `composition_demo.launch.py <https://github.com/ros2/demos/blob/master/composition/launch/composition_demo.launch.py>`_.

2. Add the package containing the server to your ``package.xml`` file.

    .. code-block:: xml

        <exec_depend>nav2_route_server</exec_depend>

Error codes 
***********

> Your nav2 task server may also wish to return a 'error_code' in its action response (though not required). If there are semantically meaningful and actionable types of failures for your system, this is a systemic way to communicate those failures which may be automatically aggregated into the responses of the navigation system to your application.

Reviewer: An empty line after a header is required

+---------------------------------------------------+-----------------------+----------------------+
| `Controller Server`_                              | NONE=0, UNKNOWN=100   | 101-199              |
+---------------------------------------------------+-----------------------+----------------------+
| `Planner Server`_ (compute_path_to_pose)          | NONE=0, UNKNOWN=200   | 201-299              |
+---------------------------------------------------+-----------------------+----------------------+
| `Planner Server`_ (compute_path_through_poses)    | NONE=0, UNKNOWN=300   | 301-399              |
+---------------------------------------------------+-----------------------+----------------------+
| ...                                               | ...                   |                      |
+---------------------------------------------------+-----------------------+----------------------+
| Last Nav2 server                                  | NONE=0, UNKNOWN=9900  | 9901-9999            |
+---------------------------------------------------+-----------------------+----------------------+

> | Last Nav2 server                                  | NONE=0, UNKNOWN=10000 | 10001-10099          |

Reviewer: Route Server?

+---------------------------------------------------+-----------------------+----------------------+
| ...                                               | ...                   |                      |
+---------------------------------------------------+-----------------------+----------------------+

.. _Controller Server: https://github.com/ros-planning/navigation2/blob/main/nav2_controller/src/controller_server.cpp
.. _Planner Server: https://github.com/ros-planning/navigation2/blob/main/nav2_planner/src/planner_server.cpp

> Error codes are attached to the response of the action message. An example can be seen below for the route server

Reviewer: ```suggestion
Error codes are attached to the response of the action message. An example can be seen below for the route server. Note that by convention we set the error code field within the message definition to ``error_code``.

```

.. code-block:: bash

    # Error codes
    # Note: The expected priority order of the errors should match the message order
    uint16 NONE=0 # 0 is reserved for NONE
    uint16 UNKNOWN=10000 # first error code in the sequence is reserved for UNKNOWN

    # User Error codes below
    int16 INVILAD_START=10001
    int16 NO_VALID_ROUTE=10002

    #goal definition
    route_msgs/PoseStamped goal
    route_msgs/PoseStamped start
    string route_id
    ---
    #result definition
    nav_msgs/Route route
    builtin_interfaces/Duration route_time
    uint16 error_code
    ---

As stated in the message, the priority order of the errors should match the message order, 0 is reserved for NONE and the first error code in the sequence is reserved for UNKNOWN.
Since the the route server is a external server, the errors codes start at 10000 and go up to 10099.

In order to propigate your server's error code to the rest of the system it must be added to the nav2_params.yaml file. 

> The `error_code_id_names` define what error codes to compare. The lowest error code of the sequence is then returned. 

Reviewer: ```suggestion
The `error_code_id_names` define what error codes to compare. The lowest error code of the sequence is then returned - whereas the code enums increase the higher up in the software stack - giving higher priority to lower-level failures.

```

.. _lifecycle_composition:

Adding a New Nav2 Task Server 
#############################

> A nav2 task server consists of server side logic to complete different types of requests, usually called by the autonomy system or through the Behavior Tree Navigator. In this guide, we will discuss the core components needed to add a new task server to Nav2 (ex. Controller, Behavior, Smoother, Planner Servers). Namely, how to set up your new Lifecycle-Component Node for launch and state management and the communication of semantically meaningful error codes (if necessary). 

Reviewer: ```suggestion
A nav2 task server consists of server side logic to complete different types of requests, usually called by the autonomy system or through the Behavior Tree Navigator. In this guide, we will discuss the core components needed to add a new task server to Nav2 (ex. Controller, Behavior, Smoother, Planner Servers). Namely, how to set up your new Lifecycle-Component Node for launch and state management and the communication of semantically meaningful error codes (if necessary) to a behavior tree or application to be fully integrated into the framework. 

```

.. code-block:: bash

    # Error codes
    # Note: The expected priority order of the errors should match the message order
    uint16 NONE=0 # 0 is reserved for NONE
    uint16 UNKNOWN=10000 # first error code in the sequence is reserved for UNKNOWN

    # User Error codes below
    int16 INVILAD_START=10001
    int16 NO_VALID_ROUTE=10002

    #goal definition
    route_msgs/PoseStamped goal
    route_msgs/PoseStamped start
    string route_id
    ---
    #result definition
    nav_msgs/Route route
    builtin_interfaces/Duration route_time
    uint16 error_code
    ---

As stated in the message, the priority order of the errors should match the message order, 0 is reserved for NONE and the first error code in the sequence is reserved for UNKNOWN.
Since the the route server is a external server, the errors codes start at 10000 and go up to 10099.

In order to propigate your server's error code to the rest of the system it must be added to the nav2_params.yaml file. 

> The `error_code_id_names` define what error codes to compare. The lowest error code of the sequence is then returned - whereas the code enums increase the higher up in the software stack - giving higher priority to lower-level failures.

Reviewer: ```suggestion
The `error_code_id_names` inside of the BT Navigator define what error codes to look for on the blackboard by the server. The lowest error code of the sequence is then returned - whereas the code enums increase the higher up in the software stack - giving higher priority to lower-level failures.

```
